_posts/2020-06-24-trainable-segmentation.markdown

```diff
@@ -0,0 +1,106 @@
+
+In machine learning, a sample is represented as a vector of *features*. Deep
+learning models learn features directly from the data and are very popular for
+image processing. Nevertheless, they require a large training set and their
+training is very resource-intensive. We use instead local features representing
+for each pixel
+- the average intensity in a small region around the pixel
+- the average magnitude of gradients in the same region
+- and measures of local texture in this region
+
+Such features are computed by first convolving the image of interest with a Gaussian
+kernel, and then measuring the local intensity, gradient intensity, or the
+eigenvalues of the Hessian matrix. Conveniently, these operations are provided
+by the [`filters` module of `scikit-image`](https://scikit-image.org/docs/stable/api/skimage.filters.html) and are relatively fast, since they
+operate on local neighbourhoods. 
+
+```python
+import numpy as np
+from skimage import filters
+
+def _singlescale_basic_features(img, sigma, intensity=True, edges=True,
+                                texture=True):
+    """Features for a single value of the Gaussian blurring parameter ``sigma``
+    """
+    features = []
+    img_blur = filters.gaussian(img, sigma)
+    if intensity:
+        features.append(img_blur)
+    if edges:
+        features.append(filters.sobel(img_blur))
+    if texture:
+        H_elems = [
+            np.gradient(np.gradient(img_blur)[ax0], axis=ax1)
+            for ax0, ax1 in combinations_with_replacement(range(img.ndim), 2)
+        ]
+        eigvals = feature.hessian_matrix_eigvals(H_elems)
+        for eigval_mat in eigvals:
+            features.append(eigval_mat)
+    return features
+``` 
+
+These local features can be computed for each color channel of the image and
+for different scales `sigma`. Large `sigmas` are useful to capture variations
+characteristic of textures but they will make it harder to classify pixels
+lying close to the boundary between objects. Users can modify the set of
+features in a control panel thanks to interactive elements from
+[dash-core-components](https://dash.plotly.com/dash-core-components): 
+a checklist for the type of features and a range
+slider for the sigma parameter. 
+
+![screenshot of app showing image and control panel](/assets/trainable_segmentation_panel.png)
+
+In order to build the training set, we use the [new shape drawing capabilities
+of plotly.py](https://eoss-image-processing.github.io/jekyll/update/2020/05/06/shape-drawing.html)
+and in particular the `drawopenpath` dragmode which can used to draw
+"squiggles" on parts of the image which you want to label. In order to optimize
+the annotation time, the width of the squiggle is adjusted with a Dash
+`dcc.Slider`. Each time a new annotation is drawn, it is [captured thanks to a
+relayoutData event of the plotly figure which triggers a callback](https://dash.plotly.com/interactive-graphing).
```

Comment:
At first reading it sounds like the width of the stroke influences the computation time, but you mean that it makes annotation faster by being able to cover more of the image with a wider stroke? How about:
```suggestion
"squiggles" on parts of the image which you want to label. The width of the squiggle can be adjusted with a Dash
`dcc.Slider` to make it possible to annotate features of different sizes. Each time a new annotation is drawn, it is [captured by the plotly figure's
relayoutData event, which triggers a callback]](https://dash.plotly.com/interactive-graphing).
```
(also just changed the wording a little bit to give it "anglo-saxon austereness" ;))